src/server/schedulers/initMaterializedViews.ts

```diff
@@ -0,0 +1,81 @@
+import { Job, Queue, Worker } from 'bullmq'
+import IORedis from 'ioredis'
+import { Objects } from 'utils/objects'
+import { Promises } from 'utils/promises'
+
+import { CountryIso } from 'meta/area'
+
+import { AreaController } from 'server/controller/area'
+import { AssessmentController } from 'server/controller/assessment'
+import { BaseProtocol, DB } from 'server/db'
+import { CountryActivityLogRepository } from 'server/repository/assessmentCycle/countryActivityLog'
+import { Logger } from 'server/utils/logger'
+
+const client: BaseProtocol = DB
+
+// assessmentUuid->cycleUuid->[countryIso]
+type ActivityLogEntries = Record<string, Record<string, Array<CountryIso>>>
+
+const _getActivityLogEntries = async (props: { jobTimestamp: number }): Promise<ActivityLogEntries> => {
+  const { jobTimestamp } = props
+
+  const jobCreatedAt = new Date(jobTimestamp).toISOString()
+
+  const activities = await client.map<{ assessmentUuid: string; cycleUuid: string; countryIso: CountryIso }>(
+    `select assessment_uuid, cycle_uuid, country_iso from public.activity_log where time > $1 order by time desc`,
+    [jobCreatedAt],
+    (row) => Objects.camelize(row)
+  )
+  return activities.reduce<ActivityLogEntries>((acc, entry) => {
+    const res = { ...acc }
+    if (!acc[entry.assessmentUuid]) res[entry.assessmentUuid] = {}
+    if (!acc[entry.assessmentUuid][entry.cycleUuid]) res[entry.assessmentUuid][entry.cycleUuid] = []
+    res[entry.assessmentUuid][entry.cycleUuid].push(entry.countryIso)
+    return acc
+  }, {})
+}
+
+export const initMaterializedViews = (connection: IORedis): Worker => {
+  const name = 'Scheduler-MaterializedViews'
+  const queue = new Queue<void>(name, { connection, streams: { events: { maxLen: 1 } } })
+
+  const worker = new Worker(
+    name,
+    async (job: Job) => {
+      Logger.info(`[${name}] ** started`)
+
+      const [assessments, entries] = await Promise.all([
+        AssessmentController.getAll({}),
+        _getActivityLogEntries({ jobTimestamp: job.timestamp }),
+      ])
+
+      await Promises.each(assessments, (assessment) =>
+        Promises.each(assessment.cycles, async (cycle) => {
+          // 1. refresh country summary
+          await AreaController.refreshSummaries({ assessment, cycle })
+          Logger.info(`[${name}:CountrySummary] ${assessment.props.name} ${cycle.name} refreshed`)
+
+          // 2. refresh countries activity log
+          const countryISOs = entries[assessment.uuid]?.[cycle.uuid] ?? []
+          await Promises.each(countryISOs, async (countryIso) => {
+            Logger.debug(`[${name}:CountryActivityLog] ${assessment.props.name} ${cycle.name} refreshing`)
+            await CountryActivityLogRepository.refreshMaterializedView({ assessment, cycle, countryIso })
+            Logger.info(`[${name}:CountryActivityLog] ${assessment.props.name} ${cycle.name} refreshed`)
```

Comment:
Separate function for 1 and 2 for readability?

eg

```suggestion
_refreshSummer(assessment, cycle)
_refreshCountryActivityLog(countryISOs, assessment, cycle)
```

Reply:
Open the file , it's not that bad . We can do that in the future if needed . What do you think ?